Smartplaylist only updates playlists that may have changed

brunal opened the pull request:
`database_changed` sends the model that changed (stored/deleted/updated) and the `smartplaylist` plugin uses that information to detect which playlists it should regenerate.
Query usage has also been improved: the plugin uses `OrQuery`, which is way faster that merging individual queries.

This is a work in progress:
- [x] there seems to be problems on query building: Kings of Leon matches my French playlist according to smartplaylist, but `ls` disagrees.
- [x] sort is not managed. There does not seem to be a right way to deal with it.
- [x] there are only skeletons of tests
- [x] the update to `database_changed` should be documented
- [x] maybe a let's-find-what-this-function-needs approach in the function wrapper (beets/plugins.py:114) should be restored (@sampsyo I need your input there).

edit: motivation for that change: playlist building can be really slow (1min+), even if I barely did anything. This solves it.

sampsyo commented on beets/dbcore/query.py:
```diff
@@ -177,8 +189,8 @@ class RegexpQuery(StringFieldQuery):
     expression.
     """
-    def __init__(self, field, pattern, false=True):
-        super(RegexpQuery, self).__init__(field, pattern, false)
```
Wow, `false=True`. That's some Orwellian shit.

sampsyo commented:
Looks great! And hashable queries seem useful for other purposes.

What do you mean by "a let's-find-what-this-function-needs approach"? Happy to help, as usual.

brunal commented:
Here is sort management! I slightly modified Sort parsing to that no useless MultipleSort() is instantiated (see commit 45c0c9b).

> What do you mean by "a let's-find-what-this-function-needs approach"? Happy to help, as usual.

I'm referring to the previous behaviour that offered compatibility to any listener that did not accept all arguments by inspecting its arguments.

sampsyo commented:
Aha, great—sorry for being slow.

Sure, I'd be happy with reinstating that. It's definitely hacky, but as you've discovered here, it's nice to be able to change events in minor ways without breaking the API contract.

brunal commented:
Here is backwards-compatibility on import stages and event listeners! I added tests as well.

I think that this PR is now complete.

sampsyo commented:
Looks great to me! Hitting the merge button. :sparkles:

djl commented:
Has the API significantly changed since this patch was applied? I've starting getting a traceback when importing a new album:

    Traceback (most recent call last):
      File "/Users/djl/Projects/beets/beet", line 20, in <module>
        beets.ui.main()
      File "/Users/djl/Projects/beets/beets/ui/__init__.py", line 1081, in main
        _raw_main(args)
      File "/Users/djl/Projects/beets/beets/ui/__init__.py", line 1071, in _raw_main
        subcommand.func(lib, suboptions, subargs)
      File "/Users/djl/Projects/beets/beets/ui/commands.py", line 892, in import_func
        import_files(lib, paths, query)
      File "/Users/djl/Projects/beets/beets/ui/commands.py", line 869, in import_files
        session.run()
      File "/Users/djl/Projects/beets/beets/importer.py", line 316, in run
        pl.run_parallel(QUEUE_SIZE)
      File "/Users/djl/Projects/beets/beets/util/pipeline.py", line 301, in run
        out = self.coro.send(msg)
      File "/Users/djl/Projects/beets/beets/util/pipeline.py", line 183, in coro
        func(*(args + (task,)))
      File "/Users/djl/Projects/beets/beets/importer.py", line 1331, in apply_choices
        task.add(session.lib)
      File "/Users/djl/Projects/beets/beets/importer.py", line 670, in add
        self.album = lib.add_album(self.imported_items())
      File "/Users/djl/Projects/beets/beets/library.py", line 1187, in add_album
        item.add(self)
      File "/Users/djl/Projects/beets/beets/library.py", line 280, in add
        super(LibModel, self).add(lib)
      File "/Users/djl/Projects/beets/beets/dbcore/db.py", line 428, in add
        self.store()
      File "/Users/djl/Projects/beets/beets/library.py", line 273, in store
        plugins.send('database_change', lib=self._db, model=self)
      File "/Users/djl/Projects/beets/beets/plugins.py", line 458, in send
        result = handler(**arguments)
      File "/Users/djl/Projects/beets/beets/plugins.py", line 126, in wrapper
        return func(*args, **kwargs)
      File "/Users/djl/Projects/beets/beetsplug/smartplaylist.py", line 138, in db_change
        matches = q.match(model) or q.match(model.get_album())
      File "/Users/djl/Projects/beets/beets/dbcore/query.py", line 412, in match
        return all([q.match(item) for q in self.subqueries])
      File "/Users/djl/Projects/beets/beets/dbcore/query.py", line 113, in match
        return self.value_match(self.pattern, item.get(self.field))
      File "/Users/djl/Projects/beets/beets/dbcore/query.py", line 110, in value_match
        raise NotImplementedError()
    NotImplementedError

Here's my config:

    smartplaylist:
      playlists:
        - name: "000-latest.m3u"
          query: added:@{1m}.. albumartist+ album+ track+
        - name: "$year.m3u"
          query: year:@{,%Y} albumartist+ album+ track+


The weird `@{}` syntax is from my [custom plugin](https://github.com/djl/dotfiles/blob/0bf81e7dda8151d2ad4803b51d681018ea96477e/.config/beets/plugins/timedelta.py) which does a simple string replace in the query (the plugin is documented in the docstring if you need to know how it works). Yes, this is quite hacky but it's the only way to easily manipulate a query since template functions [are not supported in queries](https://github.com/sampsyo/beets/issues/1147).

If I disable my plugin (or change the query to not include my syntax) I get no traceback so it's clearly coming from my plugin, I just don't know where.

Also this only happens when importing. Running `beet splupdate` manually still works fine.

Should I open a new issue for this?

sampsyo commented:
Thanks for reporting. I don't quite see why this worked before and not now, but I do see the solution for your plugin: add a definition for `value_match` in your FieldQuery subclass. This can just forward to `self._query`, I think, but it needs to be defined for the smartplaylist plugin to do its one-at-a-time matching, as distinct from wholesale SQLite querying.

djl commented:
That won't work either :(

`self._query` will be an instance of `DateQuery` or `NumericQuery` neither of which define a `value_match` method.

sampsyo commented:
Ah, right you are. No problem: you can define `match` instead (you need one or the other, not both) and forward *that* to the sub-query.

djl commented:
Perfect! Not sure how I misssed that :expressionless:

Thanks!

brunal commented:
I've had the same issue as @djl while importing. I'll have a good look at it tomorrow evening/on Monday.

brunal commented:
I just pushed a fix on master. Since in the end I was not able to reproduce the problem I'm not 100% sure I've fixed it though.... I've also added `repr()` to queries for easier future debugging.
